### ISSUE TITLE
PacBio Kinnex data

### ISSUE THREAD

**Issue by cjfields** [benjjneb/dada2#1892]
We're just getting in our first standard 16S data from the PacBio Revio which is using the Kinnex kit, and will soon be getting 16S-ITS-23S data.  I know the FASTQ data are binned now (similar to NovaSeq I believe).  Is anyone aware of issues with these?  I'm more concerned about issues similar to the NovaSeq binning issues documented elsewhere:

* #791 
* #1307 

I'm also wondering whether the kit itself (which concatenates the reads prior to sequencing, and then splits the resulting reads after sequencing based on PacBio adapters) may also affect the error profile.

**Comment by cjfields** [comment]
It does appear there are issues with `learnErrors`, at least when using our workflow.  We run this script:

```{r}
filts <- list.files('.', pattern="filtered.fastq.gz", full.names = TRUE)
sample.namesF <- sapply(strsplit(basename(filts), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
set.seed(100)

# Learn forward error rates
errs <- learnErrors(filts, 
    errorEstimationFunction=PacBioErrfun, 
    BAND_SIZE=32, 
    multithread=4)

pdf(paste0("R1",".err.pdf"))
plotErrors(errs, nominalQ=TRUE)
dev.off()
```

The error appears to be due to the binning.  From our Nextflow output

```
Command exit status:
  140

Command output:
  213565386 total bases in 147764 reads from 1 samples will be used for learning the error rates.                                   

Command error:
  The max qual score of 93 was not detected. Using standard error fitting.                                                          
  The max qual score of 93 was not detected. Using standard error fitting.                                                          
  The max qual score of 93 was not detected. Using standard error fitting.  
```

We're working through the workflow using R+dada2 and a small data set to manually diagnose it and see if we can create a workaround.

**Comment by cjfields** [comment]
It looks like the issue above was a workflow problem (the task timed out).  However we are seeing increased loss of sequence data compared to past runs, and there are similar issues with the binning step as seen for the NovaSeq runs interestingly enough.  The error profile also looks like an issue, similar to the dip seen in NovaSeq runs.

Attached are the results from three random samples from a larger 174 sample run using `plotQualityProfile` (you can ignore the 'aggregate' labels, these are from individual samples).   Lots of read data per sample.  I'll see if I can dig up some comparative standard Sequel IIe runs.  

[aggregate-qualities.12.pdf](https://github.com/benjjneb/dada2/files/14365611/aggregate-qualities.12.pdf)
[aggregate-qualities.27.pdf](https://github.com/benjjneb/dada2/files/14365612/aggregate-qualities.27.pdf)
[aggregate-qualities.77.pdf](https://github.com/benjjneb/dada2/files/14365613/aggregate-qualities.77.pdf)

This is the error profile:

![image](https://github.com/benjjneb/dada2/assets/49265/e4f95cf5-51b7-4936-b5d7-e6815e9f2c0c)

Note the dip.

**Comment by cjfields** [comment]
We just tried a run using the alternative function for error estimation provided by @jonalim and tested by @hhollandmoritz here: https://github.com/benjjneb/dada2/issues/1307#issuecomment-957680971 (option 4).  This seems to help, though the data are pretty sparse; we're working on newer runs with more `nbases` to see if this improves anything

<img width="700" alt="image" src="https://github.com/benjjneb/dada2/assets/49265/ef8d7e5d-1f28-444f-931a-9aa8418f89a3">

**Comment by cjfields** [comment]
This is an updated plot with `nbases=1e10`

![image](https://github.com/benjjneb/dada2/assets/49265/0d3024f9-b70d-4cff-9da9-40d1d65e3e70)

I should mention that the PacBio workflow steps for Kinnex still appear to use the original `PacBioErrFunc`.  We are downloading their Zymo mock samples to see if the pattern above holds:

https://downloads.pacbcloud.com/public/dataset/Kinnex-16S/DATA-ZymoATCCMix-32plex-Revio/2-Demux/fastq_files/

I should mention these data are from the Revio, I'll update the title.

**Comment by cjfields** [comment]
Error profiles for the mock community samples from here:

https://downloads.pacbcloud.com/public/dataset/Kinnex-16S/DATA-ZymoATCCMix-32plex-Revio/2-Demux/fastq_files/

## The original PacBio model, `PacBiofunc`

![image](https://github.com/benjjneb/dada2/assets/49265/5f917b92-ee47-484c-98f0-0566e890d1f6)

## Updated version

Using the code from https://github.com/benjjneb/dada2/issues/1307#issuecomment-957680971 (option 4):

![image](https://github.com/benjjneb/dada2/assets/49265/c0634a6e-83aa-49c7-9cb8-df84e1b50ed7)

At the moment we're going with the second one.  @benjjneb any thoughts on this?

**Comment by benjjneb** [comment]
Visually the fit with the updated version looks much better. Without a deep understanding, my surface read is that the updated code might be more effectively working across Q scores with very low representation in the data, and thus the improved visual fitting could be explained by a better match between the algorithm and this sort of binned-Q data.

We are actually looking at PacBio Kinnex/Revio for a current project so this is very timely. I haven't ever seen real data from the platform before, but googling around I just found this: https://downloads.pacbcloud.com/public/dataset/Kinnex-16S/

**Comment by cjfields** [comment]
> Visually the fit with the updated version looks much better. Without a deep understanding, my surface read is that the updated code might be more effectively working across Q scores with very low representation in the data, and thus the improved visual fitting could be explained by a better match between the algorithm and this sort of binned-Q data.

Thx, but the original implementation was from the _Illumina_ NovaSeq binning tests from @jonalim (which I think came from someone in his group...).  That this works across binned data from different technologies is interesting on it's own.

> We are actually looking at PacBio Kinnex/Revio for a current project so this is very timely. I haven't ever seen real data from the platform before, but googling around I just found this: https://downloads.pacbcloud.com/public/dataset/Kinnex-16S/

Yep, that's part of the data I had tried here: 

https://github.com/benjjneb/dada2/issues/1892#issuecomment-1961647289

32-sample Zymo mock communities.  Similar pattern when we ran it.  I noticed they also have Kinnex/Sequel data.

**Comment by cjfields** [comment]
Small update: based on feedback from our PacBio BI rep it sounds like SMRTLink can be configured to generate full quality scores for Revio, though it sounds like these top out at Q40 regardless (unlike Sequel with the Q93).  It requires a custom run sheet for now.  Default is to bin all data.

**Comment by cjfields** [comment]
@benjjneb to add to this, we redid a Kinnex run but set it to generate the full quality range.  Despite PacBio's indication these would top out at 40 they do go all the way to Q93 (using the Kinnex correction above):

<img width="670" alt="image" src="https://github.com/benjjneb/dada2/assets/49265/4c88eb58-81cd-425f-b8c6-7a5db68bbbd7">

We're doing a second run through switching back to the standard PacBio model, which may work better in this case.

**Comment by cjfields** [comment]
As a final addition, if you use the full quality scores the original PacBio error function in DADA2 it seems to work well:

<img width="707" alt="image" src="https://github.com/benjjneb/dada2/assets/49265/0673cc06-e375-4a39-931e-7d8cd66c2968">

Note that the qualities go all the way to 93 again.  However, I noticed that the `dada()` step takes a tremendous amount of time, approximately 3-4 days using 36 cores and 300GB mem with a full SMRTcell and ~200 samples.

**Comment by jonalim** [comment]
I appreciate this thread a whole lot @cjfields ! I had also noticed the reduced quality range on our first Revio Kinnex run. Unrelated to this, our flowcell was underloaded(sic?) and the median read quality less than what it should have been, so I was confused as to what was going on.

I'm actually unconvinced that the error model a lot of people have been attributing to me (with loess `weights = log10(tot), span=0.95, degree = 1`) is better than the original, because it often yields a fit that underestimates the error rate of Q37 (illumina) / Q40 (PacBio) bases, but I'll save that for another post.

**Comment by cjfields** [comment]
@jonalim agreed, and I think this is where having a decent mock data set would be useful to assess this.  There are ones available from PacBio which we have tested and which I linked to above.

**Comment by cjfields** [comment]
We'll likely move to the new error model developed for binned quality scores with these data as well, though it will need testing:

https://github.com/benjjneb/dada2/issues/1307#issuecomment-2521790524

**Comment by splaisan** [comment]
Hi,
I am using the nice nextflow developped at pacbio by @proteinosome and now wonder how much the use of dada2 in that pipeline is affected by the biased qscores provided with the kinex reads. Can you please comment on the consequences of the error model bias.
Thanks

**Comment by benjjneb** [comment]
> I am using the nice nextflow developped at pacbio by [@proteinosome](https://github.com/proteinosome) and now wonder how much the use of dada2 in that pipeline is affected by the biased qscores provided with the kinex reads. Can you please comment on the consequences of the error model bias.

I don't think the quality scores are biased, they are just binned into a much smaller number of values than the 0-93 that one previously got with HiFi or the 0-40 that was the norm on other sequencing technologies. This does cause some issues with fitting the error model using the standard method, although evaluations suggest this usually isn't that deleterious to the denoising process.

That said, we think a binned-quality score model is more appopriate, and have added one such model to the latest devel version of DADA2. However, since Pacbio's Nextflow pipeline is using the DADA2 QIIME2 plugin, I'm not sure how long it will take for this error model to be available there. First it has to make it into the Bioconductor release of DADA2 R package, then added to the QIIME2 plugin and make it into a QIIME2 release, and then be added into the Pacbio Nextflow workflow.

**Comment by splaisan** [comment]
Thanks a lot for sharing, if the Pacbio devs read this and are interested I am sure they will look into it.

**Comment by splaisan** [comment]
Thank you @proteinosome , knowing you know about it is enough for me, no hurry!
I recommend your tool to our sequencing customers and my only requirement is that it be doing the job correctly and trustfully.
Please continue providing us straightforward tools like this one.

**Comment by splaisan** [comment]
Hi @cjfields, 
If this can help, I regularly run 200-300 samples pools and it takes few hours indeed on 80 threads but is much faster if I limit each sample to a max of 20k or 40k reads to start with using subsample=20000 in my config.
I often have a number of samples in a library with >200k reads which are 'abusing' the pipeline). I do not think much more than 20-40k is necessary to explore a standard sample. Of course saturation will give you a better value but often the balance between low and very high sample sequencing depth makes it difficult to find a compromise.

**Comment by cjfields** [comment]
@splaisan missed your last comment! I agree in some respects though I'd say "it largely depends on the project" (a common refrain for every bio project I've worked on for the last 20 yrs). We regularly work with complex microbiome data, many from very diverse sources (sea water and sediment, rumen, soil, etc) and sometimes mixed within the same sequencing run. Alpha rarefaction analyses on these samples suggest subsampling can lose informative lower-abundance ASVs and affect richness estimates, something reported elsewhere (e.g., [here](https://pubmed.ncbi.nlm.nih.gov/35070172/)). But those may not matter as much if filtering data for beta diversity and differential abundance analysis.

In short: it really depends on the project goals and the samples.

**Comment by momsane** [comment]
Hi,
first of all thanks a lot for putting so much effort into adding features to dada2.
I am piggybacking on this issue since I also just received some Kinnex data from a Revio run. The samples that were sequenced contain a simple defined community of 25 strains (with a few rare contaminants). I have been looking at all the discussions regarding Revio data and to me it is still not clear which error model should be used. I am not at all an expert like other contributors in this discussion but there are some things that don't make sense to me. 

In the fastQ files the quality scores are indeed capped at 40, but looking at the quality plots from the sequencing facility and from fastQC they don't seem to be binned. This was further confirmed by our sequencing platform. I first tried to use the standard `PacBioErrfun` and got the same kind of plots as other people (all also the message about Q93 not detected).

[04_error_plot_PacBioErrfun.pdf](https://github.com/user-attachments/files/20625842/04_error_plot_PacBioErrfun.pdf)

However I noticed that the points that were well aligned on the error rates lines often (but not systematically) corresponded to the score bins reported [here](https://ccs.how/faq/qv-binning.html). So then I tried to use `makeBinnedQualErrfunc(3, 10, 17, 22, 27, 35, 40)` and got error plots where points corresponding to the bins align with the model and the rest align with the nominalQ. 

[04_error_plot_binnedQualErrfun.pdf](https://github.com/user-attachments/files/20625940/04_error_plot_binnedQualErrfun.pdf)

Finally I also tested the `loessErrfun_mod4` function from the code shared above but got the following message:
```
"689163076 total bases in 467506 reads from 231 samples will be used for learning the error rates.
Initializing error rates to maximum possible estimate.
Error rates could not be estimated (this is usually because of very few reads).
Error in getErrors(err, enforce = TRUE) : Error matrix is NULL.
```

Here is code I used (maxBases was set to 1E10):
```
dereps <- derepFastq(filtered_trimmed_reads_paths, verbose=TRUE, n = maxReads)

set.seed(42)

if (errModel == "binnedQualErrfun"){
  print("Building error model with bins [3, 10, 17, 22, 27, 35, 40]")
  binnedQs <- c(3, 10, 17, 22, 27, 35, 40)
  binnedQualErrfun <- makeBinnedQualErrfun(binnedQs)
  error_model <- learnErrors(
    dereps,
    errorEstimationFunction = binnedQualErrfun,
    nbases = maxBases,
    randomize = T,
    BAND_SIZE = 32,
    multithread = T,
    verbose = T
  )
} else if (errModel == "loessErrfun_mod4") {
  source("/Volumes/D2c/mgarcia/20240708_mgarcia_syncom_invivo/exp01_inoculation_methods/pacbio_analysis/workflow/scripts/loessErrfun_mod4.R")
  print("Building error model with custom loessErrfun_mod4")
  error_model <- learnErrors(
    dereps,
    errorEstimationFunction = loessErrfun_mod4,
    nbases = maxBases,
    randomize = T,
    BAND_SIZE = 32,
    multithread = T,
    verbose = T
  )
} else if (errModel == "PacBioErrfun") {
  print("Building error model with standard PacBio model")
  error_model <- learnErrors(
    dereps,
    errorEstimationFunction = PacBioErrfun,
    nbases = maxBases,
    randomize = T,
    BAND_SIZE = 32,
    multithread = T,
    verbose = T
  )
}

err.plot <- plotErrors(error_model, nominalQ=TRUE)

ggsave(file.path(out.plots, paste0("04_error_plot_", errModel, ".pdf")), err.plot, device="pdf", width = 10, height = 8)
```

So it looks like none of these functions are working as expected, unless there is something I didn't catch.
